Public/Src/Utilities/Storage/FileContentInfo.cs

```diff
@@ -18,27 +18,153 @@ namespace BuildXL.Storage
+                {
+                    Contract.Assert(false, $"Invalid length value (reserved bits should not be used): {length} ({System.Text.RegularExpressions.Regex.Replace(Convert.ToString(length, 2).PadLeft(64, '0'), ".{4}", "$0 ")})");
+                }
+                
+                m_data = (existence.HasValue ? (IsKnownExistenceFlag | ((long)existence << ExistenceShift)) : 0L)
+                    | IsKnownLengthFlag
+                    | length;
+            }
+
+            /// <summary>
+            /// Creates a <see cref="LengthAndExistence"/> with an unknown length and optional existence value.
+            /// </summary>            
+            public LengthAndExistence(PathExistence? existence)
+            {
+                m_data = existence.HasValue ? (IsKnownExistenceFlag | ((long)existence << ExistenceShift)) : 0L;
+            }
+
+
+            private LengthAndExistence(long combinedValue)
+            {
+                m_data = combinedValue;
+            }
+
+            /// <summary>
+            /// Creates a <see cref="LengthAndExistence"/> using a combined existence/length value. Mainly used in deserialization logic.
+            /// </summary>                        
+            public static LengthAndExistence CreateWithCombinedValue(long combinedValue)
+            {
+                // check that if a flag is not set, the corresponding value is not set as well
+                if ((combinedValue & IsKnownExistenceFlag) == 0)
+                {
+                    Contract.Assert((combinedValue & ExistenceValueMask) == 0);
+                }
+
+                if ((combinedValue & IsKnownLengthFlag) == 0)
+                {
+                    Contract.Assert((combinedValue & LengthValueMask) == 0);
+                }
+
+                return new LengthAndExistence(combinedValue);
+            }
+        }
 
         /// <summary>
         /// Creates a <see cref="FileContentInfo"/> with real USN version information.
         /// </summary>
-        public FileContentInfo(ContentHash hash, long length)
+        public FileContentInfo(ContentHash hash, long length) 
         {
-            m_length = length;
+            Contract.Requires(length >= 0);
+            
+            m_lengthAndExistence = new LengthAndExistence(            
+                length,
+                // if the length is valid, assign the existence value
+                IsValidLength(length, hash) 
+                    ? PathExistence.ExistsAsFile 
+                    : (PathExistence?)null);            
+
+            Hash = hash;
+        }
+
+        /// <summary>
+        /// Creates a <see cref="FileContentInfo"/> using a combined existence/length value. 
+        /// </summary>
+        /// <remarks>
+        /// Mainly used for deserialization and files with unknown length.
+        /// </remarks>
+        public FileContentInfo(ContentHash hash, LengthAndExistence lengthAndExistence)
+        {

```

Comment:
Please add some tests for these to make sure things round trip correctly through the bit packing of the underlying struct